src/Features/Core/Portable/PullMemberUp/MembersPuller.cs

```diff
@@ -0,0 +1,408 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.  
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.PullMemberUp;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+using EditorMap = System.Collections.Immutable.ImmutableDictionary<Microsoft.CodeAnalysis.SyntaxTree, Microsoft.CodeAnalysis.Editing.DocumentEditor>;
+using SyntaxMap = System.Collections.Immutable.ImmutableDictionary<Microsoft.CodeAnalysis.ISymbol, Microsoft.CodeAnalysis.SyntaxNode[]>;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.PullMemberUp
+{
+    internal class MembersPuller
+    {
+        internal static readonly MembersPuller Instance = new MembersPuller();
```

Comment:
Does this have any state whatsoever? Why a singleton over just making the class static?

Reply:
This class is stateless. I just think it is not like a common helpers so choose to make it singleton.

Reply:
Note: i actually don't think this should be static class.  But i also don't think it should be singleton.  I think ti should be an instance.  That way you *could* actually store state.  It would likely make things simpler as you could pass less data across all your methods.  but that's up to you if you want to do this.